### DISCUSSION TITLE
Can recording proxy mode record an inner ssh session?

### DISCUSSION THREAD

**Discussion by alwaysastudent** [gravitational/teleport#31549] [Q&A] [answered]
Let us say I am using the recording proxy mode to get from my local to host A. And then I am using a pubkey auth from host A to host B. Will Teleport proxy be able to record what I do on host B?

local (tsh ssh)-> teleport-proxy -> host-A (ssh) -> host-B

**Comment by webvictim** [comment, marked as answer]
If you SSH directly into host B from host A after connecting to host A with Teleport, then yes. Your `ssh` command and all the subsequent commands will be captured by the session recording, as it happens in the same terminal that's already being recorded.

**Reply by alwaysastudent** [reply]
@webvictim - thank you for answering.

But can you please explain how it can record the inner session from host-a to host-b as there is no proxy involved after I get to host-a? 

local (tsh ssh)-> teleport-proxy -> host-A (ssh) -> host-B

As you see host-a to host-b is not over the teleport proxy, so my understanding was that the ssh client on "host-a" will negotiate versions, perform key exchanges, and establish an entirely new SSH session to the target "host-b". The SSH session keys are exchanged directly between the client(host-a) and the target server(host-b) here, and the encryption is supposed to be end to end without any mitm to intercept this inner ssh session. 

My understanding obviously seems to be wrong, so I would like to be schooled, thank you :)

**Reply by webvictim** [reply]
Teleport records all terminal output from and keyboard input to your session on host A. In effect, it's pointing a video camera at your screen and keyboard for the duration of your session and recording everything that happens.

If you run `ssh host-b` during that session, Teleport records it because it's still the **same** terminal being used - `ssh` is just a subprocess running that Teleport can "see" you using. Everything you type into the terminal gets relayed to the remote ssh session you're now connected to (which Teleport records) and everything that's echoed from the remote session is relayed back to your terminal (which Teleport records)

The encryption is still end to end between host A and host B - it's just that someone (Teleport) is looking over your shoulder watching the whole thing.

I hope this makes sense!

**Reply by alwaysastudent** [reply]
@webvictim thank you for the clarity and being kind. I still have a couple of questions on how Teleport Proxy can record ssh traffic certain scenarios as below.  Hope you can help me here.

1) Can Teleport Proxy record traffic over a reverse tunnel setup? Say someone ssh into "host-A" and sets up a reverse tunnel from that host into local, for port 22, and uses that tunnel from local from then on?
2) Can Teleport Proxy record traffic that is happening over something like sshuttle? https://github.com/sshuttle/sshuttle ; https://dev.to/mrkaran/using-sshtunnel-2h24
3) Can Teleport Proxy record, file transfer protocols like x/y/zmodem over SSH?

**Reply by webvictim** [reply]
1. No, if the traffic isn't visible in the terminal Teleport won't record it.
2. No, for the same reason
3. No, for the same reason

It's just stuff which is visible as being typed in a terminal which is visible. Teleport records an audit event if you set up a port forward with it (like `tsh -L 2022:server2:22 server1` as in your first example) but if you then `ssh -p 2022 localhost` that traffic is invisible to Teleport.
